### ISSUE TITLE
Preventing Implicit Import of dbsync and rsync Libraries in Windows

### ISSUE THREAD

**Issue by Dwordcito** [wazuh/wazuh#17882]
### Description

This issue involves avoiding the implicit import of the `dbsync` and `rsync` libraries in Windows. To achieve this, the static linking of the `fimdb` library needs to be avoided, allowing it to have the same behavior and handling as the `syscollector.dll` library. Additionally, the installer needs to be modified to ensure the deployment of the `fimdb.dll` library.

### Proposed Solution

The proposed solution is to modify the build process and installer to ensure the proper handling and deployment of the `fimdb.dll` library. This may involve changes in the linking configuration, build scripts, and installer settings.

### Tasks

- Review the current build process and linker configuration to identify the dependencies and their linking behavior.
- Modify the build process to avoid static linking of the `fimdb` library.
- Update the installer to include the deployment of the `fimdb.dll` library.
- Test the changes on Windows to ensure that the implicit import of `dbsync` and `rsync` libraries is prevented.
- Validate the behavior and handling of the `fimdb.dll` library in comparison to `syscollector.dll`.

### Expected Outcome

After implementing the proposed solution, the implicit import of the `dbsync` and `rsync` libraries should be prevented in Windows. The `fimdb.dll` library should follow the same behavior and handling as the `syscollector.dll` library. The installer should correctly deploy the `fimdb.dll` library alongside the other required components.

This issue needs aims to https://github.com/wazuh/wazuh/tree/4.6.0

**Comment by Dwordcito** [comment]
@pereyra-m enable this Action, after applying the proposed solution.

https://github.com/wazuh/wazuh/actions/workflows/windows-dll-hijack.yml

**Comment by pereyra-m** [comment]
## Update

After trying different approaches, the following items were completed:
- The `fimdb.dll` library now is shared and not static
- All the required exports were added
- The direct calls to rsync/dbsync in the Windows agent main method were moved
- Both .msi and .exe installers were updated with the new `libfimdb.dll` file
- The Makefile file was updated with the required changes to get a successful compilation: some code objects were removed, some include paths were added, the order of the linking libraries was fixed, etc.

So the `rsync/dbsync` implicit load was replaced with the load of `libfimdb.dll`. 
It's still required to delay the load this library and remove the `libgcc_s_dw2-1.dll` dependency

![2023-07-13_00-47](https://github.com/wazuh/wazuh/assets/65046601/1989cd92-25e8-4775-9043-3decd0cc8d98)

**Comment by pereyra-m** [comment]
### Update

The following items were completed:
- The `libfimdb.dll` library now has delayed load version and this dependency has been removed

Still it's required to:
- Test the new `libfimdb.lib` and make sure the MinGW bug isn't affecting the Wazuh binary
- Remove the `libgcc_s_dw2-1.dll`. It seems that some arithmetic operations are generating inner methods calls
https://gcc.gnu.org/onlinedocs/gccint/Integer-library-routines.html#Integer-library-routines

![2023-07-14_01-11](https://github.com/wazuh/wazuh/assets/65046601/0f9a1d81-1a24-490a-a3ed-d2c71a81085d)

**Comment by pereyra-m** [comment]
### Update

The dependency of the library `libgcc_s_dw2-1.dll` couldn't be avoided.
After testing a minimal `.exe` file compiled with MinGW, the frame register/deregister methods are always implicitly required. In other words, there isn't any piece of code in `wazuh-syscheck` that could be modified/removed to avoid this dependency.

So, the next option was to generate a `libwazuh-syscheck.dll` library that could be called from the Wazuh agent main in the entry point `Start_win32_Syscheck()`. But this code is currently a static library, and it relies on many symbols that are being added directly in the final `wazuh-agent.exe` file. Generating this library would imply adding many object files in the syscheck's CMakeLists.txt.

Also, the library `libgcc_s_dw2-1.dll` can't be statically linked because it could generate failures during some exceptions generated in non DWARF compatible frames.

The last resource is to create a delayed load version of this library. But like we aren't generating this library, a tool like `gendef` is required (included in MinGW).
Here we have the resulting `.def` file

<details><summary>Details</summary>
<p>

```
;
; Definition file of libgcc_s_dw2-1.dll
; Automatic generated by gendef
; written by Kai Tietz 2008
;
LIBRARY "libgcc_s_dw2-1.dll"
EXPORTS
_Unwind_Backtrace
_Unwind_DeleteException
_Unwind_FindEnclosingFunction
_Unwind_Find_FDE
_Unwind_ForcedUnwind
_Unwind_GetCFA
_Unwind_GetDataRelBase
_Unwind_GetGR
_Unwind_GetIP
_Unwind_GetIPInfo
_Unwind_GetLanguageSpecificData
_Unwind_GetRegionStart
_Unwind_GetTextRelBase
_Unwind_RaiseException
_Unwind_Resume
_Unwind_Resume_or_Rethrow
_Unwind_SetGR
_Unwind_SetIP
__absvdi2
__absvsi2
__addtf3
__addvdi3
__addvsi3
__ashldi3
__ashrdi3
__bswapdi2
__bswapsi2
__clear_cache
__clrsbdi2
__clrsbsi2
__clzdi2
__clzsi2
__cmpdi2
__ctzdi2
__ctzsi2
__deregister_frame
__deregister_frame_info
__deregister_frame_info_bases
__divdc3
__divdi3
__divmoddi4
__divsc3
__divtc3
__divtf3
__divxc3
__emutls_get_address
__emutls_register_common
__enable_execute_stack
__eqtf2
__extenddftf2
__extendsftf2
__extendxftf2
__ffsdi2
__ffssi2
__fixdfdi
__fixsfdi
__fixtfdi
__fixtfsi
__fixunsdfdi
__fixunsdfsi
__fixunssfdi
__fixunssfsi
__fixunstfdi
__fixunstfsi
__fixunsxfdi
__fixunsxfsi
__fixxfdi
__floatdidf
__floatdisf
__floatditf
__floatdixf
__floatsitf
__floatundidf
__floatundisf
__floatunditf
__floatundixf
__floatunsitf
__gcc_personality_v0
__getf2
__gttf2
__letf2
__lshrdi3
__lttf2
__moddi3
__muldc3
__muldi3
__mulsc3
__multc3
__multf3
__mulvdi3
__mulvsi3
__mulxc3
__negdi2
__negtf2
__negvdi2
__negvsi2
__netf2
__paritydi2
__paritysi2
__popcountdi2
__popcountsi2
__powidf2
__powisf2
__powitf2
__powixf2
__register_frame
__register_frame_info
__register_frame_info_bases
__register_frame_info_table ; Check!!! forwards to pthread_mutex_unlock in libwinpthread-1.dll (ordinal 77)
__register_frame_info_table_bases ; Check!!! forwards to pthread_mutex_unlock in libwinpthread-1.dll (ordinal 77)
__register_frame_table ; Check!!! forwards to pthread_mutex_unlock in libwinpthread-1.dll (ordinal 77)
__subtf3
__subvdi3
__subvsi3
__trunctfdf2
__trunctfsf2
__trunctfxf2
__ucmpdi2
__udivdi3
__udivmoddi4
__umoddi3
__unordtf2

```

</p>
</details>  

Now, many tests have to be performed in order to make sure that everything is working as expected after these changes.

![2023-07-17_23-51](https://github.com/wazuh/wazuh/assets/65046601/7585e2c8-d924-4f16-ba99-4d884c5c5b03)

**Comment by pereyra-m** [comment]
## Update

Finally, the `libgcc_s_dw2-1.dll` dependency can be avoided by compiling the Windows agent with gcc and not with g++.
This way, the `gendef` dependency is avoided.
See the PR for more details.

**Comment by pereyra-m** [comment]
## Update

The issue is blocked until the issue for branch maintenance fixes the state of the integration tests.
This issue https://github.com/wazuh/wazuh-qa/issues/4325 was merged for 4.5.0 but not for 4.6.0.

**Comment by pereyra-m** [comment]
## Update

Now the issue is blocked by https://github.com/wazuh/wazuh-qa/issues/4356
